### PR TITLE
Task/add grpc status request

### PR DESCRIPTION
**Pull request by badisa**

### CODE REVIEW

File: examples/hif2a/fit_to_multiple_rbfes.py [hunk 1]
```diff
@@ -155,6 +155,7 @@ def _save_forcefield(filename, ff_params):
     else:
         # Setup GRPC client
         client = GRPCClient(hosts=args.hosts)
+    client.verify()
```
Comment (badisa) [inline examples/hif2a/fit_to_multiple_rbfes.py:158]: @maxentile Added you to the PR so you were aware of the (potential) change to adding a verify method to the client

File: parallel/client.py [hunk 2]
```diff
@@ -57,6 +57,17 @@ def submit(self, task_fn, *args):
...
         raise NotImplementedError()
 
+    def verify(self):
+        """Verify performs any necessary checks to verify the client is ready to
+        handle calls to submit.
+
+        Raises
+        ------
+        Exception
+            If verfication fails
```
Comment (badisa) [inline parallel/client.py:67]: typo: verification*

File: proto/parallel/service.proto [hunk 3]
```diff
@@ -6,8 +6,19 @@ syntax = "proto3";
...
 message PickleData {
     bytes binary = 1;
+}
+
+// Incase we want to send data in the future
+message StatusRequest{
+
+}
+
+message StatusResponse {
```
Comment (badisa) [inline proto/parallel/service.proto:21]: @proteneer Let me know if anything else jumps out as important to verify.

File: parallel/client.py [hunk 4]
```diff
@@ -186,3 +187,16 @@ def submit(self, task_fn, *args):
         future = self.stubs[self._idx].Submit.future(request)
         self._idx = (self._idx + 1) % len(self.stubs)
         return BinaryFutureWrapper(future)
+
+    def verify(self):
```
Comment (badisa) [inline parallel/client.py:191]: Currently this does does not verify the host that is running the client. This seemed helpful so if you wanted to modify your local 'analysis' code while still using a homogenous set of workers to generate data.

File: parallel/client.py [hunk 5]
```diff
@@ -89,6 +100,12 @@ def submit(self, task_fn, *args):
         self._idx = (self._idx + 1) % self.max_workers
         return future
 
+    def verify(self):
+        """
+        See abstract class for documentation.
+        """
+        return
```
Comment (proteneer) [inline parallel/client.py:107]: Better to raise `NotImplementedError` here instead.

File: parallel/client.py [hunk 6]
```diff
@@ -164,3 +187,16 @@ def submit(self, task_fn, *args):
...
+        """
+        See abstract class for documentation.
+        """
+        futures = [stub.Status.future(service_pb2.StatusRequest()) for stub in self.stubs]
+        workers_status = [x.result() for x in futures]
+        for field in ("nvidia_driver", "git_sha",):
+            # All fields should be the same
+            host_values = {self.hosts[i]: getattr(x, field) for i, x in enumerate(workers_status)}
+            uni_vals = set(host_values.values())
+            assert len(uni_vals) == 1, f"Not all hosts agreed for {field}: {host_values}"
```
Comment (proteneer) [inline parallel/client.py:201]: Nice!

File: parallel/client.py (hunk 4 above)
Comment (proteneer) [reply parallel/client.py:191]: This makes sense to me. Homogeneity amongst the workers is the main thing we care about.

File: proto/parallel/service.proto (hunk 3 above)
Comment (proteneer) [reply proto/parallel/service.proto:21]: I think theses two are the big ones. There's more compiler specific things we may check later on:

- check the CUDA version that was used to compile the code (nvcc --version)
- check the CUDA_ARCH used in CMake

A more stringent/all-or-nothing approach might be to just check the hash of - I *think* they should generate identical byte code if the images are built identically.

``` bash
> md5sum custom_ops.cpython-37m-x86_64-linux-gnu.so
ae4aaa3d10e850372fdc0d51906000c7  custom_ops.cpython-37m-x86_64-linux-gnu.so
```

File: parallel/client.py (hunk 5 above)
Comment (badisa) [reply parallel/client.py:107]: But doesn't that make it that we can't treat the non-abstract classes as the same? As I see it there is no validation to do here, it is 'good'.

**Review by proteneer — APPROVED** [review]
Excellent diagnostic checks to prevent frustration during run-time! Only a minor comment re: compiled code (address as you see necessary, I'd be happy with the PR as-is).

File: parallel/client.py (hunk 5 above)
Comment (badisa) [reply parallel/client.py:107]: My thought is you can do the following

```
if client_type == "process":
   client = ProcessPoolClient()
elif client_type == "cuda":
   client = CudaClient()
 elif client_type == "grpc":
   client = GrpcClient()
 client.verify()
 ```

Comment (proteneer) [reply parallel/client.py:107]: Sorry you're totally right this was a brainfart on my end - for some reason I thought the ProcessPoolClient was the abstract class! (I hardly ever use it)

File: proto/parallel/service.proto (hunk 3 above)
Comment (badisa) [reply proto/parallel/service.proto:21]: I will investigate this further separate from this. We may run into issues with running this on heterogenous hardware as well, which perhaps this indicates is a bad idea?

Comment (proteneer) [reply proto/parallel/service.proto:21]: Agreed. Let's re-visit this later (if at all).

There are very small/subtle differences that may or may not matter between different GPUs/compilers, in fact one of those difference pertain to a technically incorrect but mostly inconsequential rounding issue I'm currently trying to address in #364. In general though, we probably can't guarantee bit-wise identical results across different GPUs (part of this probably has to do with our usage of cub).

**Pull request merged**